### ISSUE TITLE
BUG: Unable to create "safe" read-only view from writeable view

### ISSUE THREAD

**Issue by BTOdell** [numpy/numpy#24973]
### Describe the issue:

If you create a view from an owned ndarray that has `writeable` set to `False`, then the view also has `writeable` set to False. Additionally, the view's `writeable` flag cannot be set back to `True` - an exception is raised.

```
v.flags.writeable = True

Traceback (most recent call last):
    ...
ValueError: cannot set WRITEABLE flag to True of this array
```

If you create a read-only view from a view that was previously writeable, then you can still set the read-only view's back to writeable without exception.

### Reproduce the code example:

```python
import numpy as np

# Create a read-write array
a = np.zeros(shape=5, dtype=int)
assert a.flags.writeable == True

# Create a read-write view (writeable can be changed back to True)
rw_view = a.view()
assert rw_view.flags.writeable == True

# Make the view read-only
rw_view.flags.writeable = False

# Writeable can be changed back to True since base array was writeable at time of view creation
rw_view.flags.writeable = True
rw_view.flags.writeable = False

# Create read-only view
r_view = rw_view.view()
assert r_view.flags.writeable == False

# Shouldn't be able to change 'writeable' back to True,
# because the view was created when 'rw_view' was read-only
r_view.flags.writeable = True  # No exception raised!
```

### Runtime information:

1.26.1
3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)]

[{'numpy_version': '1.26.1',
  'python': '3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC '
            'v.1929 64 bit (AMD64)]',
  'uname': uname_result(system='Windows', node='Desktop-ABC', release='10', version='10.0.19045', machine='AMD64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2',
                                'AVX512F',
                                'AVX512CD',
                                'AVX512_SKX'],
                      'not_found': ['AVX512_CLX', 'AVX512_CNL', 'AVX512_ICL']}},
 {'architecture': 'SkylakeX',
  'filepath': 'D:\\REDACTED\\venv\\Lib\\site-packages\\numpy.libs\\libopenblas64__v0.3.23-293-gc2f4bdbb-gcc_10_3_0-2bde3a66a51006b2b53eb373ff767a3f.dll',
  'internal_api': 'openblas',
  'num_threads': 24,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.23.dev'}]
None

**Comment by seberg** [comment]
There are some tricks to cheat NumPy, but I have to ask about the context.  In general I would say that using `arr.flags.writeable = True` is very much a "I am a grown up, I know what I am doing" operation.
I won't claim that the current behavior of when we allow it or when not is super consistent, but I think there was probably a reason for allowing it in some cases, and I am not particularly bothered by allowing it.
(I suspect you will be able to do stuff, since you can also change it to `False` after creating the views and we don't track views.)

In other words: I would lean towards just closing this, because I don't think I see it as a bug.  If you or someone else wants a concrete improvement of the scheme when we allow it, I am happy to entertain that, though.

**Comment by BTOdell** [comment]
Here is the context:

```python
def freeze(a: ndarray) -> ndarray:
    """
    Accept any numpy array and create a read-only view
    that cannot be changed back writeable.
    """
    if a.flags.owndata:
        # Create a "safe" read-only view, where the 'writeable' flag
        # cannot be changed back to True.
        if not a.flags.writeable:
            return a.view()
        a.flags.writeable = False
        read_view = a.view()
        a.flags.writeable = True
        return read_view
    # If not owned, make best effort attempt to create a read-only view.
    # BUG: Input must be owned to be able to create "safe" read-only view
    # https://github.com/numpy/numpy/issues/24973
    if a.flags.writeable:
        a = a.view()
        a.flags.writeable = False
    return a
```

The reason I need the ndarray to be "safe" from just changing the `writable` flag back to `True`, is because I'm passing this ndarray to multiple external plugin modules and I don't want it to be as easy as `a.flags.writeable = True` to make the ndarray writeable again.

I understand that I could just make defensive copies of the ndarray, but these arrays are large and I'm trying to improve performance of my application while maintaining some level of safety/security.

The current official "rule" for the `WRITEABLE` flag is this:

> WRITEABLE can only be set True if the array owns its own memory or the ultimate owner of the memory exposes a writeable buffer interface or is a string.

So as long as the owning `ndarray` has the `writeable` flag set to True at the time when the view is created, then the view (and any views created from that view) are able to set their own `writeable` flag to True.

I understand now that this isn't really a bug, it's just a very obscure feature request to improve the consistency of how read-only views are created. It seems like a reasonable "rule" could be: if the ndarray that the view was created from, had the `writeable` flag set to True when the view was created, then the view's `writeable` flag can also be set to True.

I should always be able to do this:
```python
a = # <owned ndarray or view>
a.flags.writeable = False
v = a.view()
# v is read-only locked forever
```

Maybe this could be included in NumPy 2.0?

**Comment by seberg** [comment]
There was some likely a reason for the current behavior, I expect.  NumPy prevents you from hard crashes, but beyond that has a "we are all grownups approach here".  I don't really mind tightening it, but I honestly don't believe there is much of a reason either, unless it is truly prelevant in the wild.
(There are hacks to ensure the error without a copy, I think, but...)

It just seems not bad enough to worry about to me (and yes, I don't know who might rely on it).  You made it clear you don't want anyone to write to it, if users change it without understanding the why, that is their problem.
What I would just put merge if there is a PR, is a warning that changing it to True a bad idea unless you know why it was disabled.

(And no, if you want to change it, go through a normal deprecation which we do slowly.  No need to invoke a major release.)
